### PR TITLE
Fix bug where `cf.Field.subspace` doesn't always correctly handle   global cyclic subspaces

### LINKED ISSUE #828: `Field.subspace` sometimes fails for a global cyclic subspace
At v3.16.2, `Field.subspace` sometimes fails for a global cyclic subspace:
```python
>>> import cf
>>> f = cf.example_field(0)
>>> f.subspace(X=cf.wi(-180, 180))  # OK
<CF Field: specific_humidity(latitude(5), longitude(8)) 1>
>>> f.subspace(X=cf.wi(-170, 170))  # NOT OK: should give same result as [-180, 180]
ValueError: No indices found from: X=<CF Query: (wi [-170, 170])>
```

### PR DESCRIPTION
**Pull request by davidhassell**
Fixes #828.

This PR fixes and simplifies the treatment of cyclic subspaces in `cf.mixin.fielddomain._indices`. In the process, it refactors `cf.mixin.fielddomain.anchor` by moving the business-end of that to the new `cf.DimensionCoordinate.anchor`. This new method also allows for "cell" anchoring, which we might like to use later (see #824).

### CODE REVIEW

**Review by sadielbartholomew — COMMENTED** [review]
Lots of confusing talk of wardrobes (see in-line comments :smile:) here, otherwise all good in that the fix works with good implementation and new testing coverage, but I think we have the case of a without/`wo` query to also fix up in a similar way (see my in-line comment with the code snippet example) so please check that comment and let me know what you think.

File: cf/dimensioncoordinate.py [hunk 1]
```diff
@@ -246,6 +250,150 @@ def increasing(self):
...
+                closet to cell to *value* from above (below) for
+                increasing (decreasing) coordinates.
+
+                If False (the default) then the coordinate values are
+                transformed so that the first coordinate is the closet
+                to *value* from above (below) for increasing
+                (decreasing) coordinates.
+
+            parameters: `dict`, optional
+                TODO Return a dictionary of parameters which describe the
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/dimensioncoordinate.py]: Still a TODO. Fixed in 5e5727210d551ca14abfcd6b00fce3d70c68a53b

File: cf/mixin/fielddomain.py [hunk 2]
```diff
@@ -440,50 +440,39 @@ def _indices(self, config, data_axes, ancillary_mask, kwargs):
...
-                            set_start_stop = -a
-                        else:
-                            set_start_stop = 0
+                        anchor = value.value[1]
+
+                    item = item.persist()
+                    parameters = {}
+                    item = item.anchor(anchor, parameters=parameters)
+                    n = np.roll(np.arange(size), parameters["shift"], 0)
+                    if value.operator == "wi":
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/mixin/fielddomain.py:453]: Good catch. There already is such an wo if-block, but it just doesn't handle the special case you highlight! I'll look into it.

[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/mixin/fielddomain.py:453]: Fix with tests: 228abd412f036e94453dbd736a2a34f5f3f0eb42

[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/mixin/fielddomain.py:453]: Brilliant, thanks. Marking as resolved.

**Review by sadielbartholomew — APPROVED** [review]
Thanks for addressing my feedback. I've done a final sanity check after the new commits. All good, please merge.

**Pull request merged**